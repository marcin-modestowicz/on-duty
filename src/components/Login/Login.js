// @flow
import React, { Component } from "react";
import { observable, action, computed, runInAction } from "mobx";
import { observer } from "mobx-react";
import classnames from "classnames";
import TextInput from "../TextInput";
import Button from "../Button";
import Spinner from "./spinner.svg";
import styles from "./Login.scss";

/* global SyntheticInputEvent */

type Props = {
  onLogin: (email: string, password: string) => Promise<*>
};

@observer
class Login extends Component<Props> {
  @observable email: string = "";
  @observable password: string = "";
  @observable isError: boolean = false;
  @observable isLoading: boolean = false;

  @action
  handleLogin = (event: Event) => {
    event.preventDefault();

    this.isLoading = true;

    this.props.onLogin(this.email, this.password).catch(error => {
      runInAction("user login error", () => {
        this.isError = true;
        this.isLoading = false;
      });
    });
  };

  @action
  handleEmailChange = (event: SyntheticInputEvent<*>) => {
    this.isError = false;
    this.email = event.target.value;
  };

  @action
  handlePasswordChange = (event: SyntheticInputEvent<*>) => {
    this.isError = false;
    this.password = event.target.value;
  };

  @computed
  get isFormReady(): boolean {
    return !this.isError && !!this.email.trim() && !!this.password.trim();
  }

  render() {
    return (
      <form
        className={classnames(styles.root, { [styles.error]: this.isError })}
        onSubmit={this.handleLogin}
      >
        <div className={styles.field}>
          <TextInput
            id="email"
            value={this.email}
            label="Email"
            onChange={this.handleEmailChange}
            isInvalid={this.isError}
            marginless
          />
        </div>
        <div className={styles.field}>
          <TextInput
            id="password"
            type="password"
            value={this.password}
            label="Password"
            onChange={this.handlePasswordChange}
            isInvalid={this.isError}
            marginless
          />
        </div>
        {this.isError && (
          <div className={classnames(styles.field, styles.errorMessage)}>
            Invalid credentials, please try again.
          </div>
        )}
        <div className={styles.submit}>
          <Button
            type="submit"
            disabled={!this.isFormReady || this.isLoading}
            marginless
          >
            {this.isLoading && <Spinner />}
            Login
          </Button>
        </div>
      </form>
    );
  }
}

export default Login;
