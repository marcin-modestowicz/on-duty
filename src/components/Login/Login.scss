@import '../../styles/colors.scss';

.root {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  align-items: center;

  .field {
    margin-bottom: 1em;
  }

  .submit {
    text-align: center;
  }

  .errorMessage {
    font-size: .75em;
    color: $red;
  }
}