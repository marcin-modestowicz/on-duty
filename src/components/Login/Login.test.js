// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Login from "./Login";

describe("Login component", () => {
  let login;

  beforeAll(() => {
    login = shallow(<Login onLogin={jest.fn()} />);
  });

  test("should render", () => {
    expect(toJson(login)).toMatchSnapshot();
  });
});
