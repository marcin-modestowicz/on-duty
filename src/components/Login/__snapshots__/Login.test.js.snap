// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render 1`] = `
<form
  className="root"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <TextInput
      id="email"
      isInvalid={false}
      label="Email"
      marginless={true}
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="field"
  >
    <TextInput
      id="password"
      isInvalid={false}
      label="Password"
      marginless={true}
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <div
    className="submit"
  >
    <Button
      disabled={true}
      marginless={true}
      type="submit"
    >
      Login
    </Button>
  </div>
</form>
`;
